### DISCUSSION TITLE
APC UPS not showing values

### DISCUSSION THREAD

**Discussion by YOYONL89** [nutdotnet/WinNUT-Client#193] [Q&A] [answered]
Hello,

I'm trying to get my APC Back-UPS pro 900 connected but it seems to fail to show any value's.
The following screenshot is what i see:

![NUT](https://github.com/user-attachments/assets/d4440191-5eb9-418e-8b7c-c0a9b8e03318)

These are the variables coming from the List UPS Variables:

`APC Back-UPS 900Pro (Unknown/Unknown/Unknown)
battery.charge (Description unavailable) : 100
battery.charge.low (Description unavailable) : 50
battery.charge.warning (Description unavailable) : 50
battery.date (Description unavailable) : 2001/09/25
battery.mfr.date (Description unavailable) : 2024/06/03
battery.runtime (Description unavailable) : 2300
battery.runtime.low (Description unavailable) : 300
battery.type (Description unavailable) : PbAc
battery.voltage (Description unavailable) : 27.1
battery.voltage.nominal (Description unavailable) : 24.0
device.mfr (Description unavailable) : American Power Conversion
device.model (Description unavailable) : Back-UPS RS 900G
device.serial (Description unavailable) : 3B1231X30657
device.type (Description unavailable) : ups
driver.name (Description unavailable) : usbhid-ups
driver.parameter.pollfreq (Description unavailable) : 30
driver.parameter.pollinterval (Description unavailable) : 15
driver.parameter.port (Description unavailable) : auto
driver.parameter.productid (Description unavailable) : 0002
driver.parameter.serial (Description unavailable) : 3B1231X30657
driver.parameter.synchronous (Description unavailable) : auto
driver.parameter.vendorid (Description unavailable) : 051D
driver.version (Description unavailable) : 2.8.0
driver.version.data (Description unavailable) : APC HID 0.98
driver.version.internal (Description unavailable) : 0.47
driver.version.usb (Description unavailable) : libusb-1.0.26 (API: 0x1000109)
input.sensitivity (Description unavailable) : medium
input.transfer.high (Description unavailable) : 294
input.transfer.low (Description unavailable) : 176
input.voltage (Description unavailable) : 233.0
input.voltage.nominal (Description unavailable) : 230
ups.beeper.status (Description unavailable) : disabled
ups.delay.shutdown (Description unavailable) : 20
ups.firmware (Description unavailable) : 879.L4 .I
ups.firmware.aux (Description unavailable) : L4
ups.load (Description unavailable) : 26
ups.mfr (Description unavailable) : American Power Conversion
ups.mfr.date (Description unavailable) : 2012/08/04
ups.model (Description unavailable) : Back-UPS RS 900G
ups.productid (Description unavailable) : 0002
ups.realpower.nominal (Description unavailable) : 540
ups.serial (Description unavailable) : 3B1231X30657
ups.status (Description unavailable) : OL
ups.test.result (Description unavailable) : No test initiated
ups.timer.reboot (Description unavailable) : 0
ups.timer.shutdown (Description unavailable) : -1
ups.vendorid (Description unavailable) : 051d
`
This is the debug log:
`
7-11-2024 21:32:37 [12464, WinNUT]: Battery Status => Discharging
7-11-2024 21:32:38 [12464, UPS_Device]: Enter Retrieve_UPS_Datas
7-11-2024 21:32:38 [12464, UPS_Device]: battery.charge is not supported by server.
7-11-2024 21:32:38 [12464, UPS_Device]: Apply Fallback Value when retrieving battery.charge
7-11-2024 21:32:38 [12464, UPS_Device]: battery.voltage is not supported by server.
7-11-2024 21:32:38 [12464, UPS_Device]: Apply Fallback Value when retrieving battery.voltage
7-11-2024 21:32:38 [12464, UPS_Device]: battery.runtime is not supported by server.
7-11-2024 21:32:38 [12464, UPS_Device]: Apply Fallback Value when retrieving battery.runtime
7-11-2024 21:32:38 [12464, UPS_Device]: input.frequency is not supported by server.
7-11-2024 21:32:38 [12464, UPS_Device]: Apply Fallback Value when retrieving input.frequency
7-11-2024 21:32:38 [12464, UPS_Device]: input.voltage is not supported by server.
7-11-2024 21:32:38 [12464, UPS_Device]: Apply Fallback Value when retrieving input.voltage
7-11-2024 21:32:38 [12464, UPS_Device]: output.voltage is not supported by server.
7-11-2024 21:32:38 [12464, UPS_Device]: Apply Fallback Value when retrieving output.voltage
7-11-2024 21:32:38 [12464, UPS_Device]: ups.load is not supported by server.
7-11-2024 21:32:38 [12464, UPS_Device]: Apply Fallback Value when retrieving ups.load
7-11-2024 21:32:38 [12464, UPS_Device]: Unable to calculate UPS Batt_Charge: Batt_Voltage (-1) out of range.
7-11-2024 21:32:38 [12464, UPS_Device]: Unable to calculate battery runtime, missing UPS variables.
7-11-2024 21:32:38 [12464, UPS_Device]: Output_Voltage: -1, Batt_Voltage: -1, Batt_Capacity: -1, Batt_Charge: -1
7-11-2024 21:32:38 [12464, UPS_Device]: ups.status is not supported by server.
7-11-2024 21:32:38 [12464, UPS_Device]: Apply Fallback Value when retrieving ups.status
7-11-2024 21:32:38 [12464, UPS_Device]: UPS statuses have not changed since last update, skipping.
7-11-2024 21:32:38 [12464, WinNUT]: Updating UPS data for Form.
7-11-2024 21:32:38 [12464, WinNUT]: Updating battery icons based on charge percent: -1%
7-11-2024 21:32:38 [12464, WinNUT]: Update Dial
7-11-2024 21:32:38 [12464, WinNUT]: Update Icon
7-11-2024 21:32:38 [12464, WinNUT]: NotifyIcon Text => 
WinNUT Client - 2.3
Connected
On Battery (-1%)

7-11-2024 21:32:38 [12464, WinNUT]: Battery Status => Discharging
7-11-2024 21:32:38 [12464, Pref_Gui]: User clicked ViewLog button.
7-11-2024 21:32:38 [12464, Pref_Gui]: Opened UI window to log location.
`

Can anyone tell me what i'm doing wrong?

**Comment by gbakeman** [comment]
Hi there,

Thanks for the detailed issue report. Looking at the debug log, it seems that your NUT server is not providing any variables. I'd suggest restarting it as that sometimes fixes these issues, and if not you may need to pursue further troubleshooting in that regard. I'll transfer this to a discussion for now in case you want further assistance.

**Reply by YOYONL89** [reply]
Hi gbakeman,

I will try a restart of the services although i'm curious if the UPS variable list gets populated (The option for the list is greyed out when not connected) you should assume its passing through information, especially if i get the serialnumber 540 ups.serial (Description unavailable) : 3B1231X30657 .

Also i have a webased LXC container running on my network that also gets the data from the nut server and this one seems to work quite well.
![image](https://github.com/user-attachments/assets/6e1a4e93-6abb-46f2-b748-3eddf69f7050)

Is there any extra debugging possible?

Cheers,

**Reply by gbakeman** [reply]
You make a good point there - it looks like WinNUT is getting valid data even though it's not displaying it. What's concerning to me is all of the _x is not supported by server_ lines in your log - that means WinNUT is getting `ERR VARNOTSUPPORTED` responses from the NUT server when it tries to query individual variables (`GET VAR (ups_name) (var_name)`). When you use the _List Variables_ function in WinNUT (as well as the container app you used above, I suspect), a different command is run on the NUT server (`LIST VAR (ups_name)`). I'm not ruling out an explicit bug in WinNUT yet, but I'd like for you to do some additional info gathering for me if you don't mind.

I'd like for you to open up a serial terminal session to the NUT server using your favorite client (`telnet (nut ip address):3492` for example) then run a few commands for me:

- VER
- NETVER
- LIST UPS

Those should give us some basic information about your NUT server, as well as the exact name of the UPS device(s) connected just to be certain. After that, please run:

- LIST VAR (ups name)
- GET VAR (ups name) battery.charge
- GET VAR (ups name) device.mfr
- GET VAR (ups name) input.voltage

And those commands should confirm how the NUT server responds.

If you could please provide the output for all of those commands, I'd really appreciate it. Sorry for the trouble, and thank you for your efforts on this!

**Comment by YOYONL89** [comment]
Hi gbakeman,

Rebooting everything did not resolve it.
May i point your attention to the post above?
Perhaps it gives some extra pointers or at least will answer a few questions of mine.

Cheers,

**Comment by YOYONL89** [comment, marked as answer]
Hi gbakeman,

The results of the requested querys's

VER:
`Network UPS Tools upsd 2.8.0 - http://www.networkupstools.org/`

NETVER
`1.3`

LIST UPS
`BEGIN LIST UPS
UPS apc "APC Back-UPS RS 900 G"
END LIST UPS`

LIST VAR
`BEGIN LIST UPS
UPS apc "APC Back-UPS RS 900 G"
END LIST UPS
LIST VAR APC BACK-UPS RS 900 G

BEGIN LIST VAR APC
VAR APC battery.charge "100"
VAR APC battery.charge.low "50"
VAR APC battery.charge.warning "50"
VAR APC battery.date "2001/09/25"
VAR APC battery.mfr.date "2024/06/03"
VAR APC battery.runtime "1585"
VAR APC battery.runtime.low "300"
VAR APC battery.type "PbAc"
VAR APC battery.voltage "27.3"
VAR APC battery.voltage.nominal "24.0"
VAR APC device.mfr "American Power Conversion"
VAR APC device.model "Back-UPS RS 900G"
VAR APC device.serial "3B1231X30657  "
VAR APC device.type "ups"
VAR APC driver.name "usbhid-ups"
VAR APC driver.parameter.pollfreq "30"
VAR APC driver.parameter.pollinterval "15"
VAR APC driver.parameter.port "auto"
VAR APC driver.parameter.productid "0002"
VAR APC driver.parameter.serial "3B1231X30657"
VAR APC driver.parameter.synchronous "auto"
VAR APC driver.parameter.vendorid "051D"
VAR APC driver.version "2.8.0"
VAR APC driver.version.data "APC HID 0.98"
VAR APC driver.version.internal "0.47"
VAR APC driver.version.usb "libusb-1.0.26 (API: 0x1000109)"
VAR APC input.sensitivity "medium"
VAR APC input.transfer.high "294"
VAR APC input.transfer.low "176"
VAR APC input.voltage "235.0"
VAR APC input.voltage.nominal "230"
VAR APC ups.beeper.status "disabled"
VAR APC ups.delay.shutdown "20"
VAR APC ups.firmware "879.L4 .I"
VAR APC ups.firmware.aux "L4  "
VAR APC ups.load "37"
VAR APC ups.mfr "American Power Conversion"
VAR APC ups.mfr.date "2012/08/04"
VAR APC ups.model "Back-UPS RS 900G"
VAR APC ups.productid "0002"
VAR APC ups.realpower.nominal "540"
VAR APC ups.serial "3B1231X30657  "
VAR APC ups.status "OL"
VAR APC ups.test.result "No test initiated"
VAR APC ups.timer.reboot "0"
VAR APC ups.timer.shutdown "-1"
VAR APC ups.vendorid "051d"
END LIST VAR APC
`
Now here it gets interesting as you can see NUT triggers on the APC part and not the full name.
In WinNut i did put in the full name of the UPS: APC Back-UPS RS 900 G so apparantly the full name seems to resolve the list but fails in the individual GET VAR instruction.
If i only put in APC in the name WinNut works. :O

![image](https://github.com/user-attachments/assets/3d187a13-876c-4412-bdcb-9861eac4d174)

When using the full name:

![image](https://github.com/user-attachments/assets/0a0ae431-698d-40d6-956e-c4c5260baba9)

Is this something within WinNut?

**Reply by gbakeman** [reply]
Hi,

It looks like (almost) everything is working as intended! I should've asked you for the full log and/or name given to the UPS in the NUT configuration as you'll probably see a line in the WinNUT log that says something along the lines of "invalid UPS". The NUT server assigns a name to the UPS in the configuration file, so yours is **apc** (you can see this in LIST UPS command - the first part, apc, is the name. The following bit, APC Back-UPS RS 900 G, is the description.)

What I don't understand is how WinNUT is still able to list the variables with an invalid UPS name given to it. When I try it, I see this:

![{BC96C7FC-B04E-4514-9C62-87D1FE016612}](https://github.com/user-attachments/assets/a9714270-68ad-445d-b4b1-9c883ef159a2)

Looking into it further though, I can see how WinNUT refers to variables of the UPS name slightly differently between individual calls to the GET VAR command and the LIST VAR command so that may have something to do with it. If you're interested in seeing further improvements to WinNUT handling of invalid UPS names, I've just opened an issue as #194.

Thanks again for your detailed comments here and assistance with getting to the bottom of this!

**Comment by YOYONL89** [comment]
Hi,

Sure no problem, i can try to help you with it.
I made 2 debug logs one with the correct name and one with the incorrect one:

Correct:
[WinNUT-Client-2024-11-12_OK.log](https://github.com/user-attachments/files/17720756/WinNUT-Client-2024-11-12_OK.log)

Not correct:
[WinNUT-Client-2024-11-12_NOK.log](https://github.com/user-attachments/files/17720757/WinNUT-Client-2024-11-12_NOK.log)

When thinking about it in the most simple way, it might be because of the APC part in front of the name.
Maybe thats enough in supressing the error message?
Because when i remove the APC part i get the following:
![image](https://github.com/user-attachments/assets/2b8af348-1263-4991-85bd-827c161a72a1)

Just a thought! :)

Cheers,

**Reply by gbakeman** [reply]
I think you're absolutely right - something to do with the way WinNUT is handling partial strings (or the NUT server?) seems to be making this an inconsistent problem. I'll see if I can find out more in the #194 issue. Thanks again!
